### DISCUSSION TITLE
빈 주입 방식 4가지에 대한 장단점에 대해 알려주세요

### DISCUSSION THREAD

**Discussion by 3Juhwan** [2024-woowacourse-study/level-interview#119] [Spring Core]
.

**Comment by gyungchan-jo** [comment]
1. 생성자 주입
- 장점:
    - 불변성 보장: 객체 생성 시 의존성 주입으로 불변성을 유지.
    - 안정성: NPE를 방지할 수 있다.
- 단점:
    - 복잡성 증가: 매개변수가 많아지면 생성자가 복잡해질 수 있음.

2. 수정자 주입(setter)
- 장점:
    - 유연성: 객체 생성 후에도 의존성 변경 가능.
- 단점:
    - 불완전 상태 가능성: 객체가 불완전한 상태로 존재할 수 있음.

3. 필드 주입
- 장점:
    - 간결성: 코드가 간결하고 설정이 쉬움.
- 단점:
    - 테스트 어려움: 외부에서 변경이 어려워 테스트가 어려워진다.
    - 유연성: DI 컨테이너 안에서만 작동하게 되어, DI 프레임워크가 없으면 아무것도 할 수 없다.

4. 일반 메서드 주입
- 장점:
    - 유연성: 특정 상황에서만 필요한 의존성을 주입 가능.
    - 다양한 주입: 여러 의존성을 한 번에 주입 가능.
- 단점:
    - 유지보수 어려움: 의존성이 여러 메서드에 흩어져 있을 경우 유지보수가 어려움.

**Comment by nayonsoso** [comment]
### 1. 필드 주입

@Autowired 어노테이션을 필드 위에 달아 객체를 주입하는 방식이다.
코드가 간결하지만, 명시적으로 의존성 주입을 요구하지 않기 때문에 테스트도 어렵고, 불변성을 보장하기도 어렵다.
또한 순환참조가 런타임 시점에서 발생할 수 있다.

### 2. setter 주입

setter로 원하는 의존성을 주입하는 방식이다.
객체 생성 후에도 의존성을 바꿀 수 있다는 것이 장점이자 단점이다.
런타임에 의존성을 바꿀 수 있지만, NPE 를 초래하기도 한다.

### 3. 생성자 주입

필요한 의존성을 생성자의 인자로 두어, 스프링 컨테이너가 빈을 등록할 때 생성자의 인자에 해당하는 빈을 주입받는 방식이다.
final 키워드와 함께 사용하면 null 을 방지할 수 있으며, 의존성을 불변하게 관리할 수 있다.
그리고 빈생성과 동시에 의존성 주입이 완성되기 때문에 코드의 안정성을 높일 수 있다.
명시적인 의존성 주입을 요구한다는 점 때문에 테스트를 하기도 쉽다.
스프링 4.3 부터는 클래스에 하나의 생성자만 있는 경우 @AutoWired 어노테이션을 생략할 수 있다.

### 4. 인자 주입

함수나 생성자의 인자에 직접 @AutoWired 를 붙이는 방법이다.
이 방법은 불가능한 것은 아니지만, 공식 문서에서 추천하지 않는 방식이다. 
왜냐하면 '@AutoWired 인자 주입'은 JUnit을 제외하곤 스프링 코어에서 전부 무시하는 부분이기 때문이다.
[공식문서](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/beans/factory/annotation/Autowired.html)

이 방법의 장점은 @AutoWired(required = false) 를 통해 선택적으로 의존성 주입을 받을 수 있다는 점이다.
하지만 이마저도 생성자 주입의 장점인 '의존성 명시'를 해치므로 권장되지 않는다.


### 스프링 공식의 추천

> The Spring team generally advocates constructor injection, as it lets you implement application components as immutable objects and ensures that required dependencies are not null.

일관성, 불변성의 장점 덕분에 생성자 주입 방식이 권장된다.
하지만 생성자의 인자가 너무 많아진다면 좋지 않은 코드이며 (bad code smell!)
클래스가 너무 많은 책임을 지고 있는지 확인해야 한다.
그리고 setter를 통한 주입은 필수적이지 않은 의존성에 활용 수 있으며,
디폴트 값을 설정해서 NPE를 방지할 수 있다.
